### ISSUE TITLE
Fails to build shared libraries when compiled with -DBUILD_SHARED_LIBS=ON

### ISSUE THREAD

**Issue by h3xx** [GothenburgBitFactory/taskwarrior#2410]
Attempts to build libtask.so, linking `task` to it, which fails. (It doesn't subsequently install `libtask.so`)

Suggested fix: Declare libtask and libshared to be `STATIC` (it used to state this in `src/CMakeLists.txt`).

Steps to reproduce:

    mkdir _build
    cd _build
    cmake -DBUILD_SHARED_LIBS=ON ..
    make 

<details>
<summary>Full build log</summary>

    -- The C compiler identification is GNU 10.2.0
    -- The CXX compiler identification is GNU 10.2.0
    -- Detecting C compiler ABI info
    -- Detecting C compiler ABI info - done
    -- Check for working C compiler: /usr/bin/cc - skipped
    -- Detecting C compile features
    -- Detecting C compile features - done
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Check for working CXX compiler: /usr/bin/c++ - skipped
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    -- Configuring C++17
    -- System: Linux
    -- Performing Test _HAS_CXX17
    -- Performing Test _HAS_CXX17 - Success
    -- Looking for libshared
    -- Found libshared
    -- Looking for SHA1 references
    -- Looking for GnuTLS
    -- Found GnuTLS: /usr/lib64/libgnutls.so (found version "3.6.15")
    -- Looking for timegm
    -- Looking for timegm - found
    -- Looking for get_current_dir_name
    -- Looking for get_current_dir_name - found
    -- Looking for wordexp
    -- Looking for wordexp - found
    -- Performing Test HAVE_TM_GMTOFF
    -- Performing Test HAVE_TM_GMTOFF - Success
    -- Performing Test HAVE_ST_BIRTHTIME
    -- Performing Test HAVE_ST_BIRTHTIME - Failed
    -- Looking for libuuid
    -- Looking for uuid_unparse_lower
    -- Looking for uuid_unparse_lower - found
    -- Found libuuid
    -- Configuring cmake.h
    -- Configuring man pages
    -- Configuring done
    -- Generating done
    -- Build files have been written to: /tmp/git/taskwarrior/_build
    Scanning dependencies of target columns
    Scanning dependencies of target task
    Scanning dependencies of target libshared
    Scanning dependencies of target commands
    [  1%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/Msg.cpp.o
    [  1%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/Pig.cpp.o
    [  1%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/Duration.cpp.o
    [  1%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/ip.cpp.o
    [  1%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/format.cpp.o
    [  1%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/Configuration.cpp.o
    [  3%] Building CXX object src/columns/CMakeFiles/columns.dir/ColDepends.cpp.o
    [  5%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/Color.cpp.o
    [  5%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/JSON.cpp.o
    [  6%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/FS.cpp.o
    [  8%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/Datetime.cpp.o
    [ 10%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/Timer.cpp.o
    [ 10%] Building CXX object src/columns/CMakeFiles/columns.dir/ColDescription.cpp.o
    [ 10%] Building CXX object src/columns/CMakeFiles/columns.dir/Column.cpp.o
    [ 10%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/Table.cpp.o
    [ 10%] Building CXX object src/columns/CMakeFiles/columns.dir/ColEnd.cpp.o
    [ 12%] Building CXX object src/columns/CMakeFiles/columns.dir/ColDue.cpp.o
    [ 13%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/shared.cpp.o
    [ 13%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/unicode.cpp.o
    [ 15%] Building CXX object src/columns/CMakeFiles/columns.dir/ColEntry.cpp.o
    [ 15%] Building CXX object src/columns/CMakeFiles/columns.dir/ColID.cpp.o
    [ 17%] Building CXX object src/CMakeFiles/task.dir/CLI2.cpp.o
    [ 18%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/RX.cpp.o
    [ 18%] Building CXX object src/commands/CMakeFiles/commands.dir/Command.cpp.o
    [ 20%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/utf8.cpp.o
    [ 20%] Building CXX object src/CMakeFiles/libshared.dir/libshared/src/wcwidth6.cpp.o
    [ 22%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdAdd.cpp.o
    [ 22%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdAliases.cpp.o
    [ 24%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdAnnotate.cpp.o
    [ 25%] Building CXX object src/columns/CMakeFiles/columns.dir/ColIMask.cpp.o
    [ 25%] Building CXX object src/columns/CMakeFiles/columns.dir/ColMask.cpp.o
    [ 27%] Building CXX object src/columns/CMakeFiles/columns.dir/ColModified.cpp.o
    [ 27%] Building CXX object src/columns/CMakeFiles/columns.dir/ColParent.cpp.o
    [ 29%] Building CXX object src/columns/CMakeFiles/columns.dir/ColProject.cpp.o
    [ 29%] Building CXX object src/columns/CMakeFiles/columns.dir/ColRecur.cpp.o
    [ 31%] Building CXX object src/columns/CMakeFiles/columns.dir/ColScheduled.cpp.o
    [ 31%] Building CXX object src/columns/CMakeFiles/columns.dir/ColStart.cpp.o
    [ 31%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdAppend.cpp.o
    [ 32%] Building CXX object src/columns/CMakeFiles/columns.dir/ColStatus.cpp.o
    [ 34%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdAttributes.cpp.o
    [ 34%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdBurndown.cpp.o
    [ 36%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdCalc.cpp.o
    [ 36%] Building CXX object src/columns/CMakeFiles/columns.dir/ColTags.cpp.o
    [ 36%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdCalendar.cpp.o
    [ 37%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdCommands.cpp.o
    [ 37%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdColor.cpp.o
    [ 37%] Building CXX object src/columns/CMakeFiles/columns.dir/ColTypeDate.cpp.o
    [ 37%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdColumns.cpp.o
    [ 39%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdConfig.cpp.o
    [ 39%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdContext.cpp.o
    [ 41%] Building CXX object src/columns/CMakeFiles/columns.dir/ColTypeNumeric.cpp.o
    [ 41%] Building CXX object src/columns/CMakeFiles/columns.dir/ColTypeDuration.cpp.o
    [ 43%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdCount.cpp.o
    [ 43%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdCustom.cpp.o
    [ 44%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdDelete.cpp.o
    [ 46%] Linking CXX shared library liblibshared.so
    [ 46%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdDenotate.cpp.o
    [ 48%] Building CXX object src/columns/CMakeFiles/columns.dir/ColTypeString.cpp.o
    [ 48%] Building CXX object src/columns/CMakeFiles/columns.dir/ColUntil.cpp.o
    [ 50%] Building CXX object src/columns/CMakeFiles/columns.dir/ColUrgency.cpp.o
    [ 50%] Building CXX object src/columns/CMakeFiles/columns.dir/ColUDA.cpp.o
    [ 50%] Built target libshared
    [ 51%] Building CXX object src/columns/CMakeFiles/columns.dir/ColUUID.cpp.o
    [ 53%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdDiagnostics.cpp.o
    [ 53%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdDone.cpp.o
    [ 55%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdDuplicate.cpp.o
    [ 55%] Building CXX object src/columns/CMakeFiles/columns.dir/ColWait.cpp.o
    [ 55%] Building CXX object src/CMakeFiles/task.dir/Context.cpp.o
    [ 56%] Building CXX object src/CMakeFiles/task.dir/DOM.cpp.o
    [ 56%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdEdit.cpp.o
    [ 58%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdExec.cpp.o
    [ 58%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdExport.cpp.o
    [ 60%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdGet.cpp.o
    [ 60%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdHelp.cpp.o
    [ 62%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdHistory.cpp.o
    [ 62%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdIDs.cpp.o
    [ 62%] Building CXX object src/CMakeFiles/task.dir/Eval.cpp.o
    [ 63%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdImport.cpp.o
    [ 63%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdInfo.cpp.o
    [ 65%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdLog.cpp.o
    [ 65%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdLogo.cpp.o
    [ 65%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdModify.cpp.o
    [ 67%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdProjects.cpp.o
    [ 67%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdPrepend.cpp.o
    [ 68%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdPurge.cpp.o
    [ 68%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdReports.cpp.o
    [ 70%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdShow.cpp.o
    [ 72%] Linking CXX static library libcolumns.a
    [ 72%] Built target columns
    [ 72%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdStart.cpp.o
    [ 74%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdStats.cpp.o
    [ 74%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdStop.cpp.o
    [ 75%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdSummary.cpp.o
    [ 75%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdSync.cpp.o
    [ 77%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdTags.cpp.o
    [ 77%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdTimesheet.cpp.o
    [ 79%] Building CXX object src/CMakeFiles/task.dir/Filter.cpp.o
    [ 81%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdUDAs.cpp.o
    [ 81%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdUndo.cpp.o
    [ 82%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdUnique.cpp.o
    [ 82%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdUrgency.cpp.o
    [ 84%] Building CXX object src/commands/CMakeFiles/commands.dir/CmdVersion.cpp.o
    [ 84%] Building CXX object src/CMakeFiles/task.dir/Hooks.cpp.o
    [ 86%] Building CXX object src/CMakeFiles/task.dir/Lexer.cpp.o
    [ 86%] Building CXX object src/CMakeFiles/task.dir/TDB2.cpp.o
    [ 87%] Building CXX object src/CMakeFiles/task.dir/Task.cpp.o
    [ 87%] Building CXX object src/CMakeFiles/task.dir/TLSClient.cpp.o
    [ 89%] Building CXX object src/CMakeFiles/task.dir/Variant.cpp.o
    [ 89%] Building CXX object src/CMakeFiles/task.dir/ViewTask.cpp.o
    [ 89%] Building CXX object src/CMakeFiles/task.dir/dependency.cpp.o
    [ 91%] Building CXX object src/CMakeFiles/task.dir/feedback.cpp.o
    [ 91%] Building CXX object src/CMakeFiles/task.dir/legacy.cpp.o
    /tmp/git/taskwarrior/src/Variant.cpp: In member function 'bool Variant::operator_partial(const Variant&) const':
    /tmp/git/taskwarrior/src/Variant.cpp:1113:7: warning: this statement may fall through [-Wimplicit-fallthrough=]
     1113 |       if (left.trivial () || right.trivial ())
          |       ^~
    /tmp/git/taskwarrior/src/Variant.cpp:1116:5: note: here
     1116 |     case type_boolean:
          |     ^~~~
    [ 93%] Building CXX object src/CMakeFiles/task.dir/nag.cpp.o
    [ 94%] Building CXX object src/CMakeFiles/task.dir/recur.cpp.o
    [ 94%] Building CXX object src/CMakeFiles/task.dir/rules.cpp.o
    [ 94%] Building CXX object src/CMakeFiles/task.dir/sort.cpp.o
    [ 96%] Building CXX object src/CMakeFiles/task.dir/util.cpp.o
    [ 96%] Linking CXX static library libcommands.a
    [ 96%] Built target commands
    [ 96%] Linking CXX shared library libtask.so
    [ 96%] Built target task
    Scanning dependencies of target calc_executable
    Scanning dependencies of target task_executable
    Scanning dependencies of target lex_executable
    [ 96%] Building CXX object src/CMakeFiles/calc_executable.dir/calc.cpp.o
    [ 96%] Building CXX object src/CMakeFiles/lex_executable.dir/lex.cpp.o
    [ 98%] Building CXX object src/CMakeFiles/task_executable.dir/main.cpp.o
    [100%] Linking CXX executable lex
    [100%] Linking CXX executable task
    /usr/bin/ld: libtask.so: undefined reference to `Command::read_only() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::factory(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Command*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Command*> > >&)'
    /usr/bin/ld: libtask.so: undefined reference to `Command::accepts_filter() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::needs_gc() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::accepts_modifications() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::displays_id() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::accepts_miscellaneous() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::uses_context() const'
    /usr/bin/ld: libtask.so: undefined reference to `Column::factory(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Column*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Column*> > >&)'
    /usr/bin/ld: libtask.so: undefined reference to `Command::read_only() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::factory(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Command*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Command*> > >&)'
    /usr/bin/ld: libtask.so: undefined reference to `Command::accepts_filter() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::needs_gc() const'
    collect2: error: ld returned 1 exit status
    /usr/bin/ld: libtask.so: undefined reference to `Command::accepts_modifications() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::displays_id() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::accepts_miscellaneous() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::uses_context() const'
    /usr/bin/ld: libtask.so: undefined reference to `Column::factory(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Column*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Column*> > >&)'
    make[2]: *** [src/CMakeFiles/lex_executable.dir/build.make:109: src/lex] Error 1
    make[1]: *** [CMakeFiles/Makefile2:285: src/CMakeFiles/lex_executable.dir/all] Error 2
    make[1]: *** Waiting for unfinished jobs....
    collect2: error: ld returned 1 exit status
    make[2]: *** [src/CMakeFiles/task_executable.dir/build.make:109: src/task] Error 1
    make[1]: *** [CMakeFiles/Makefile2:345: src/CMakeFiles/task_executable.dir/all] Error 2
    [100%] Linking CXX executable calc
    /usr/bin/ld: libtask.so: undefined reference to `Command::read_only() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::factory(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Command*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Command*> > >&)'
    /usr/bin/ld: libtask.so: undefined reference to `Command::accepts_filter() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::needs_gc() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::accepts_modifications() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::displays_id() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::accepts_miscellaneous() const'
    /usr/bin/ld: libtask.so: undefined reference to `Command::uses_context() const'
    /usr/bin/ld: libtask.so: undefined reference to `Column::factory(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Column*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Column*> > >&)'
    collect2: error: ld returned 1 exit status
    make[2]: *** [src/CMakeFiles/calc_executable.dir/build.make:109: src/calc] Error 1
    make[1]: *** [CMakeFiles/Makefile2:315: src/CMakeFiles/calc_executable.dir/all] Error 2
    make: *** [Makefile:171: all] Error 2

</details>

**Comment by h3xx** [comment]
HEAD is 2f47226f91f0b02f7617912175274d9eed85924f, or current master@HEAD.

**Comment by h3xx** [comment]
Okay, that solves the mystery of why I saw you had fixed it, but it wasn't included in the `master` branch.

I built from `master` thinking it was the latest.

Using version-named branches for current development locus is not the layout I was expecting this repo to have. I was expecting the `master` or a `develop` branch to be the current development locus. But if the current repo layout helps organize the team's thoughts, that's fine too.

I'll close this issue. For now, I'm just working around it by adding `-DBUILD_SHARED_LIBS=OFF` to my packaging script.
